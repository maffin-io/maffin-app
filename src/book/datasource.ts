import initSqlJs from 'sql.js';
import { DataSource } from 'typeorm';
import { useSWRConfig } from 'swr';

import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from './entities';

export async function initDB(database: Uint8Array): Promise<DataSource> {
  await initSqlJs({
    // Required to load the wasm binary asynchronously. Of course, you can host it wherever you want
    locateFile: file => `https://sql.js.org/dist/${file}`,
  });

  let datasource: DataSource;

  if (database.length !== 0) {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      database,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
  } else {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
    await createEmptyBook();
  }

  return datasource;
}

// TODO: Need to mutate the accounts SWR key so Root is shown!
async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}
