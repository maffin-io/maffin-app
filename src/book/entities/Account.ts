import * as v from 'class-validator';
import {
  Column,
  Entity, JoinColumn,
  ManyToOne,
  OneToMany,
  RelationId,
  Tree,
  TreeParent,
  TreeChildren,
  SaveOptions,
} from 'typeorm';
import type { QueryClient } from '@tanstack/react-query';

import {
  getAllowedSubAccounts,
  ASSET_ACCOUNTS,
  LIABILITY_ACCOUNTS,
} from '@/book/helpers/accountType';
import type Commodity from './Commodity';
import Split from './Split';
import BaseEntity from './BaseEntity';

/**
 * CREATE TABLE accounts (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   name            text(2048) NOT NULL,
 *   account_type    text(2048) NOT NULL,
 *   commodity_guid  CHAR(32) NOT NULL,
 *   commodity_scu   integer NOT NULL,
 *   non_std_scu     integer NOT NULL,
 *   parent_guid     CHAR(32),
 *   code            text(2048),
 *   description     text(2048),
 *   hidden          integer NOT NULL,
 *   placeholder     integer NOT NULL
 * );
 */

@Entity('accounts')
@Tree('nested-set')
export default class Account extends BaseEntity {
  static CACHE_KEY = ['api', 'accounts'];
  static TYPES = [
    'ROOT',
    'EQUITY',
    'INCOME',
    'EXPENSE',
    ...ASSET_ACCOUNTS,
    ...LIABILITY_ACCOUNTS,
  ];

  @Column({
    type: 'text',
    length: 2048,
  })
  @v.Matches(/[a-zA-Z.]+/)
  @v.Length(1, 2048)
    name!: string;

  @Column({
    type: 'text',
    enum: Account.TYPES,
    length: 2048,
    name: 'account_type',
  })
  @CheckAccountType()
  @v.IsIn(Account.TYPES)
    type!: string;

  path!: string;

  @TreeParent()
  @JoinColumn({ name: 'parent_guid' })
  @v.IsNotEmpty({ message: 'parent is required' })
  @v.ValidateIf(o => o.type !== 'ROOT')
  // Seems TreeParent doesn't let us create referencing the guid of the account
  // and needs the whole object...
    parent!: Account;

  @RelationId((account: Account) => account.parent)
    parentId: string;

  @TreeChildren()
    children!: Account[];

  @RelationId((account: Account) => account.children)
    childrenIds: string[];

  @ManyToOne('Commodity', { eager: true, cascade: true })
  @JoinColumn({ name: 'commodity_guid' })
  @CheckIECommodity()
  @v.ValidateIf(o => o.type !== 'ROOT')
  @v.IsNotEmpty({ message: 'commodity is required' })
    fk_commodity!: Commodity | string;

  get commodity(): Commodity {
    return this.fk_commodity as Commodity;
  }

  @OneToMany('Split', (split: Split) => split.fk_account)
    splits!: Split[];

  @Column({
    type: 'text',
    length: 2048,
    nullable: true,
  })
  @v.IsOptional()
  @v.Length(4, 2048)
    description?: string;

  @Column({
    default: false,
  })
    hidden!: boolean;

  @CheckPlaceholder()
  @Column({
    default: false,
  })
    // Placeholders are hierarchical accounts that don't containg
    // transactions, they are just parents of other accounts
    placeholder!: boolean;

  async save(options?: SaveOptions): Promise<this> {
    const account = await super.save(options);

    if (this.queryClient) {
      updateCache({ queryClient: this.queryClient });
    }

    return account;
  }

  async remove(options?: SaveOptions): Promise<this> {
    if (this.queryClient) {
      updateCache({ queryClient: this.queryClient });
    }

    const account = await super.remove(options);
    return account;
  }
}

export async function updateCache({ queryClient }: { queryClient: QueryClient }) {
  queryClient.invalidateQueries({
    queryKey: [...Account.CACHE_KEY],
  });
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Account, 'name', { value: 'Account' });

/**
 * Checks allowed types for parent account given the current account's type
 */
function CheckAccountType(validationOptions?: v.ValidationOptions) {
  return function f(object: Account, propertyName: string) {
    v.registerDecorator({
      name: 'checkAccountType',
      target: object.constructor,
      propertyName,
      options: validationOptions,
      validator: {
        validate(type: string, args: v.ValidationArguments) {
          const account = args.object as Account;
          if (account.parent) {
            return getAllowedSubAccounts(account.parent.type).includes(type);
          }

          return true;
        },

        defaultMessage(args: v.ValidationArguments) {
          const account = args.object as Account;
          const allowedTypes = getAllowedSubAccounts(account.parent.type);

          return `only ${allowedTypes} types can be selected with ${account.parent.type} account as parent`;
        },
      },
    });
  };
}

/**
 * Checks that income/expense commodity is the same as the parent.
 *
 * This is a way to check that we always keep the same currency for income/expense
 * account as we only support them in the main currency.
 */
function CheckIECommodity(validationOptions?: v.ValidationOptions) {
  return function f(object: Account, propertyName: string) {
    v.registerDecorator({
      name: 'checkIECommodity',
      target: object.constructor,
      propertyName,
      options: validationOptions,
      validator: {
        validate(fk_commodity: Commodity, args: v.ValidationArguments) {
          const account = args.object as Account;
          if (
            account.parent
            && account.parent.type !== 'ROOT'
            && ['INCOME', 'EXPENSE'].includes(account.type)
          ) {
            return fk_commodity.guid === (account.parent.fk_commodity as Commodity).guid;
          }

          return true;
        },

        defaultMessage(args: v.ValidationArguments) {
          const account = args.object as Account;
          return `Income and Expense accounts must have ${(account.parent.fk_commodity as Commodity).mnemonic} as their commodity`;
        },
      },
    });
  };
}

/**
 * Checks if an account can be a placeholder. Accounts can be Placeholders
 * if they don't have any transactions associated.
 */
function CheckPlaceholder(validationOptions?: v.ValidationOptions) {
  return function f(object: Account, propertyName: string) {
    v.registerDecorator({
      name: 'checkPlaceholder',
      target: object.constructor,
      propertyName,
      options: validationOptions,
      validator: {
        async validate(_: string, args: v.ValidationArguments) {
          const account = args.object as Account;
          if (account.guid) {
            const [, numSplits] = await Split.findAndCount({
              where: { fk_account: { guid: account.guid } },
            });

            if (numSplits > 0 && account.placeholder) {
              return false;
            }
          }

          return true;
        },

        defaultMessage() {
          return 'Placeholder accounts cannot have transactions';
        },
      },
    });
  };
}
