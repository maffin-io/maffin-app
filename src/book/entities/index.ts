export { default as Account } from './Account';
export { default as BankConfig } from './BankConfig';
export { default as Commodity } from './Commodity';
export { default as Price } from './Price';
export { default as Split } from './Split';
export { default as Transaction } from './Transaction';
export { default as Book } from './Book';
export { default as BaseEntity } from './BaseEntity';
