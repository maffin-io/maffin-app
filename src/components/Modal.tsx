import React from 'react';

export type ModalProps = {
  open: boolean;
  setOpen: Function;
  title?: string;
  children: React.ReactNode;
};

export default function Modal({
  open, setOpen, title, children,
}: ModalProps): JSX.Element {
  if (!open) {
    return <span />;
  }

  return (
    <div
      className="fixed inset-0 bg-white bg-opacity-30 overflow-y-auto h-full w-full z-50"
      id="my-modal"
    >
      <div
        className="relative top-20 mx-auto p-5 w-1/3 shadow-lg rounded-md bg-gunmetal-700"
      >
        <button
          type="button"
          className="float-right"
          onClick={() => setOpen(false)}
        >
          X
        </button>
        {
          title
          && (
            <div className="w-full pb-4 border-b border-gunmetal-600">
              <span className="text-lg">
                {title}
              </span>
            </div>
          )
        }
        {children}
      </div>
    </div>
  );
}
