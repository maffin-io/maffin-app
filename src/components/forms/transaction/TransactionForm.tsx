import React from 'react';
import { DateTime } from 'luxon';
import { useForm } from 'react-hook-form';
import { classValidatorResolver } from '@hookform/resolvers/class-validator';
import { mutate } from 'swr';
import { IsNull } from 'typeorm';
import { Tooltip } from 'react-tooltip';

import {
  Split,
  Transaction,
} from '@/book/entities';
import { isInvestment } from '@/book/helpers/accountType';
import SplitsField from './SplitsField';
import type { FormValues } from './types';

const resolver = classValidatorResolver(Transaction, { validator: { stopAtFirstError: true } });

export type TransactionFormProps = {
  action?: 'add' | 'update' | 'delete',
  onSave: Function,
  defaultValues?: FormValues,
};

export default function TransactionForm({
  action = 'add',
  onSave,
  defaultValues,
}: TransactionFormProps): JSX.Element {
  const form = useForm<FormValues>({
    defaultValues,
    mode: 'onChange',
    resolver,
  });

  const { errors } = form.formState;

  let disabled = false;
  let submitButtonText = 'Save';
  let submitButtonClass = 'btn-primary';
  if (action === 'update') {
    submitButtonText = 'Update';
    submitButtonClass = 'btn-warn';
  }
  if (action === 'delete') {
    disabled = true;
    submitButtonText = 'Delete';
    submitButtonClass = 'btn-danger';
  }

  return (
    <form onSubmit={form.handleSubmit((data) => onSubmit(data, action, onSave))}>
      <fieldset className="text-sm my-5">
        <label htmlFor="dateInput" className="inline-block mb-2">Date</label>
        <input
          id="dateInput"
          disabled={disabled}
          className="block w-full m-0"
          {...form.register('date')}
          type="date"
        />
        <p className="invalid-feedback">{errors.date?.message}</p>
      </fieldset>

      <fieldset className="text-sm my-5">
        <label htmlFor="descriptionInput" className="inline-block mb-2">Description</label>
        <span
          className="badge"
          data-tooltip-id="description-help"
        >
          ?
        </span>
        <Tooltip
          id="description-help"
          className="bg-cyan-600 w-1/3 text-white rounded-lg p-2"
          disableStyleInjection
        >
          <p className="mb-2">
            Add a meaningful description to your transaction. The description is shown in the
            list of transactions in the account detail page.
          </p>
          <p>
            It helps to set the same description
            for similar transactions. For example, for a Groceries transaction, you can set
            &quot;Supermarket&quot; to all of them.
          </p>
        </Tooltip>
        <input
          id="descriptionInput"
          disabled={disabled}
          className="block w-full m-0"
          {...form.register('description')}
          type="text"
          placeholder="Enter description"
        />
        <p className="invalid-feedback">{errors.description?.message}</p>
      </fieldset>

      <fieldset className="text-sm my-5">
        <label className="inline-block mb-2">Splits</label>
        <SplitsField
          disabled={disabled}
          form={form}
        />
      </fieldset>

      <input
        {...form.register(
          'guid',
        )}
        hidden
        type="text"
      />

      <input
        {...form.register(
          'fk_currency',
        )}
        hidden
        type="number"
      />

      <div className="flex w-full gap-2 items-center justify-center">
        <button className={submitButtonClass} type="submit" disabled={Object.keys(errors).length > 0}>
          {submitButtonText}
        </button>
      </div>
    </form>
  );
}

async function onSubmit(data: FormValues, action: 'add' | 'update' | 'delete', onSave: Function) {
  const transaction = Transaction.create({
    ...data,
    guid: data.guid || undefined,
    date: DateTime.fromISO(data.date),
  });

  if (action === 'add') {
    await transaction.save();
  } else if (action === 'update') {
    await transaction.save();
    await Split.delete({
      fk_transaction: IsNull(),
    });
  } else if (action === 'delete') {
    // Not using cascade here because seems it has problems with
    // old data. Deleting splits manually for now
    await Split.remove(transaction.splits);
    await transaction.remove();
  }

  transaction.splits.forEach(split => {
    if (isInvestment(split.account)) {
      mutate('/api/investments');
    }
    mutate(`/api/splits/${split.account.guid}`);
  });

  mutate('/api/txs/latest');
  onSave();
}
