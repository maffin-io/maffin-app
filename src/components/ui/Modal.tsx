import React from 'react';
import RModal from 'react-modal';

export type ModalProps = {
  className?: string;
  triggerContent: string | JSX.Element;
  triggerClassName?: string;
  showClose?: boolean;
} & React.PropsWithChildren;

export interface ModalRef {
  openModal: () => void;
  closeModal: () => void;
}

function Modal({
  className = 'modal',
  triggerClassName = '',
  triggerContent,
  showClose = false,
  children,
}: ModalProps, ref: React.Ref<ModalRef>): JSX.Element {
  const [isOpen, setIsOpen] = React.useState<boolean>(false);

  // Expose actions to the parent
  React.useImperativeHandle(ref, () => ({
    openModal: () => setIsOpen(true),
    closeModal: () => setIsOpen(false),
  }));

  return (
    <>
      <RModal
        isOpen={isOpen}
        overlayClassName="overlay"
        className={className}
        onRequestClose={() => setIsOpen(false)}
      >
        {
          showClose
          && (
            <button
              type="button"
              className="float-right"
              onClick={() => setIsOpen(false)}
            >
              X
            </button>
          )
        }
        {children}
      </RModal>

      <button
        type="button"
        onClick={() => setIsOpen(!isOpen)}
        className={triggerClassName}
      >
        {triggerContent}
      </button>
    </>
  );
}

export default React.forwardRef(Modal);
