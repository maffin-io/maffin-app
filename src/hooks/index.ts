export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
