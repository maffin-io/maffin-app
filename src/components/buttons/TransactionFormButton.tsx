import React from 'react';
import { BiPlusCircle } from 'react-icons/bi';

import { DataSourceContext } from '@/hooks';
import TransactionForm from '@/components/forms/transaction/TransactionForm';
import Modal from '@/components/Modal';
import type { FormValues } from '@/components/forms/transaction/types';
import {
  Account,
  Commodity,
  Split,
  Transaction,
} from '@/book/entities';

export type TransactionFormButtonProps = {
  action?: 'add' | 'update' | 'delete',
  guid?: string, // The transaction to update or delete
  account?: Account, // Account to populate for the main split
  defaultValues?: FormValues,
  className?: string,
  children?: React.ReactNode,
};

export default function TransactionFormButton(
  {
    action = 'add',
    guid,
    account,
    defaultValues,
    children,
    className = 'btn-primary',
  }: TransactionFormButtonProps,
): JSX.Element {
  const { save } = React.useContext(DataSourceContext);
  const [isModalOpen, setIsModalOpen] = React.useState<boolean>(false);
  const [defaults, setDefaults] = React.useState(defaultValues);

  let title = 'Add transaction';
  if (action === 'update') {
    title = 'Edit transaction';
  }
  if (action === 'delete') {
    title = 'Confirm you want to remove this transaction';
  }

  return (
    <>
      <Modal
        title={title}
        open={isModalOpen}
        setOpen={setIsModalOpen}
      >
        <TransactionForm
          action={action}
          onSave={() => {
            save();
            setIsModalOpen(false);
          }}
          defaultValues={defaults}
        />
      </Modal>
      <button
        type="button"
        className={className}
        onClick={async () => {
          if (action === 'add') {
            const split1 = new Split();
            if (account) {
              split1.fk_account = account;
            }

            const split2 = new Split();
            setDefaults({
              ...defaultValues,
              splits: [split1, split2],
            } as FormValues);
          } else if (action === 'update' || action === 'delete') {
            const tx = await Transaction.findOneOrFail({
              where: { guid },
              relations: {
                splits: {
                  fk_account: true,
                },
              },
            });
            setDefaults({
              ...tx,
              date: tx.date.toISODate() as string,
              fk_currency: tx.currency as Commodity,
            });
          }
          setIsModalOpen(!isModalOpen);
        }}
      >
        {
          children
          || (
            <>
              <BiPlusCircle className="mr-1" />
              <span>Add Transaction</span>
            </>
          )
        }
      </button>
    </>
  );
}
