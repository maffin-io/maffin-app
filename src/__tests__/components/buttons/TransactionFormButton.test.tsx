import React from 'react';
import {
  act,
  render,
  screen,
  fireEvent,
  waitFor,
} from '@testing-library/react';
import Modal from 'react-modal';

import { Account, Commodity, Transaction } from '@/book/entities';
import TransactionFormButton from '@/components/buttons/TransactionFormButton';
import TransactionForm from '@/components/forms/transaction/TransactionForm';
import { DataSourceContext } from '@/hooks';
import type { DataSourceContextType } from '@/hooks';
import { DateTime } from 'luxon';

jest.mock('react-modal', () => jest.fn(
  (props: React.PropsWithChildren) => (
    <div data-testid="Modal">
      {props.children}
    </div>
  ),
));

jest.mock('@/components/forms/transaction/TransactionForm', () => jest.fn(
  () => <div data-testid="TransactionForm" />,
).mockName('TransactionForm'));
const TransactionFormMock = TransactionForm as jest.MockedFunction<typeof TransactionForm>;

describe('TransactionFormButton', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders as expected by default', async () => {
    const { container } = render(
      <TransactionFormButton />,
    );

    expect(container).toMatchSnapshot();
  });

  it('can pass children to render alternative button text', async () => {
    render(
      <TransactionFormButton>
        <span>Update</span>
      </TransactionFormButton>,
    );

    screen.getByText('Update');
  });

  it.each([
    'add',
    'update',
    'delete',
  ])('renders hidden modal with TransactionForm on mount with action %s', async (action) => {
    render(<TransactionFormButton action={action as 'add' | 'update' | 'delete'} />);
    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: false,
      }),
      {},
    );

    expect(TransactionForm).toHaveBeenLastCalledWith(
      {
        action,
        defaultValues: undefined,
        onSave: expect.any(Function),
      },
      {},
    );
  });

  it('opens modal when clicking the button', async () => {
    render(<TransactionFormButton />);

    const button = await screen.findByRole('button', { name: /add transaction/i });
    fireEvent.click(button);

    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: true,
      }),
      {},
    );
  });

  it('closes modal when clicking the X button', async () => {
    render(
      <TransactionFormButton />,
    );

    const button = await screen.findByRole('button', { name: /add transaction/i });
    fireEvent.click(button);

    await screen.findByTestId('Modal');
    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: true,
      }),
      {},
    );

    const xButton = screen.getByRole('button', { name: 'X' });
    fireEvent.click(xButton);

    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: false,
      }),
      {},
    );
  });

  it.each(
    ['update', 'delete'],
  )('retrieves original Transaction when %s', async (action) => {
    jest.spyOn(Transaction, 'findOneOrFail').mockResolvedValue({
      guid: 'guid',
      date: DateTime.fromISO('2023-01-01'),
      currency: {
        mnemonic: 'EUR',
      },
    } as Transaction);
    render(
      <TransactionFormButton guid="guid" action={action as 'update' | 'delete'} />,
    );

    const button = await screen.findByRole('button', { name: /add transaction/i });
    fireEvent.click(button);

    expect(Transaction.findOneOrFail).toHaveBeenCalledWith({
      where: { guid: 'guid' },
      relations: { splits: { fk_account: true } },
    });

    await waitFor(() => {
      expect(TransactionForm).toHaveBeenLastCalledWith(
        {
          action,
          defaultValues: {
            guid: 'guid',
            date: '2023-01-01',
            currency: {
              mnemonic: 'EUR',
            },
            fk_currency: {
              mnemonic: 'EUR',
            },
          },
          onSave: expect.any(Function),
        },
        {},
      );
    });
  });

  it('passes expected data to TransactionForm', async () => {
    const mockSave = jest.fn();
    const defaultValues = {
      date: '',
      description: '',
      fk_currency: {} as Commodity,
    };

    render(
      <DataSourceContext.Provider value={{ save: mockSave as Function } as DataSourceContextType}>
        <TransactionFormButton
          account={{
            guid: '1',
          } as Account}
          defaultValues={{
            ...defaultValues,
            splits: [],
          }}
        />
      </DataSourceContext.Provider>,
    );

    // open modal to prove that onSave closes it
    const button = await screen.findByRole('button', { name: /add transaction/i });
    fireEvent.click(button);
    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: true,
      }),
      {},
    );
    expect(TransactionForm).toHaveBeenLastCalledWith(
      {
        action: 'add',
        defaultValues: {
          ...defaultValues,
          splits: [
            {
              action: '',
              guid: expect.any(String),
              fk_account: { guid: '1' },
            },
            {
              action: '',
              guid: expect.any(String),
            },
          ],
        },
        onSave: expect.any(Function),
      },
      {},
    );

    const { onSave } = TransactionFormMock.mock.calls[0][0];
    act(() => onSave());
    expect(mockSave).toBeCalledTimes(1);

    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: false,
      }),
      {},
    );
  });
});
