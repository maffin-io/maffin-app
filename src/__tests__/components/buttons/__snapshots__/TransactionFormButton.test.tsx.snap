// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionFormButton renders as expected by default 1`] = `
<div>
  <div
    data-testid="Modal"
  >
    <button
      class="float-right"
      type="button"
    >
      X
    </button>
    <span>
      Add transaction
    </span>
    <div
      data-testid="TransactionForm"
    />
  </div>
  <button
    class="btn-primary"
    id="add-tx"
    type="button"
  >
    <svg
      class="mr-1"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4z"
      />
      <path
        d="M12 2C6.486 2 2 6.486 2 12s4.486 10 10 10 10-4.486 10-10S17.514 2 12 2zm0 18c-4.411 0-8-3.589-8-8s3.589-8 8-8 8 3.589 8 8-3.589 8-8 8z"
      />
    </svg>
    <span>
      Add Transaction
    </span>
  </button>
</div>
`;
